pkg/cassandra/util.go

```diff
@@ -33,35 +33,39 @@ func DatacenterStopping(dc *cassdcapi.CassandraDatacenter) bool {
 	return dc.GetConditionStatus(cassdcapi.DatacenterStopped) == corev1.ConditionTrue && dc.Status.CassandraOperatorProgress == cassdcapi.ProgressUpdating
 }
 
-// GetDatacentersForSystemReplication determines the DCs that should be included for
-// replication. This function should only be used for system keyspaces. Replication for
-// system keyspaces is initially set through the management-api, not CQL. This allows us
-// to specify non-existent DCs for replication even though Cassandra 4 does not allow that.
-// That cannot be done when configuration replication through CQL which is why this func
-// should only be used for system keyspaces.
-func GetDatacentersForSystemReplication(kc *api.K8ssandraCluster) []api.CassandraDatacenterTemplate {
-	datacenters := make([]api.CassandraDatacenterTemplate, 0)
-	initialized := kc.Status.GetConditionStatus(api.CassandraInitialized) == corev1.ConditionTrue
-	for _, dc := range kc.Spec.Cassandra.Datacenters {
-		if dc.Stopped {
-			continue
-		}
-		if initialized {
-			status, found := kc.Status.Datacenters[dc.Meta.Name]
-			if found && status.Cassandra.GetConditionStatus(cassdcapi.DatacenterReady) == corev1.ConditionTrue {
-				datacenters = append(datacenters, dc)
+// GetDatacentersForReplication determines the DCs that should be included for replication. This function works for both
+// system and non-system keyspaces. Replication for system keyspaces is initially done before the cluster has been
+// initialized, and is set through the management-api, not CQL. This allows us to specify non-existent DCs for
+// replication even though Cassandra 4 does not allow that. Once the cluster is initialized however, replication is done
+// through CQL, and we cannot include non-existing DCs anymore; this is going to be the case: 1) when updating system
+// keyspaces; 2) when configuring keyspaces for Stargate and Reaper; and 3) when updating user keyspaces during a DC
+// rebuild.
+func GetDatacentersForReplication(kc *api.K8ssandraCluster) []api.CassandraDatacenterTemplate {
```

Comment:
I don't think combining the functions is going to work. We do the schema checks after reconciling each DC. Let's say we are creating a cluster with 3 DCs - dc1, dc2, dc3. The `CassandraInitialized` condition won't get set until all three DCs are ready, but the schema checks happen after we finish reconciling each DC (i.e., each DC is ready or stopped). With this change we will include non-existing DCs.

I did a quick test to verify. I created this cluster:

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  name: test
spec:
  cassandra:
    serverVersion: "4.0.1"
    storageConfig:
      cassandraDataVolumeClaimSpec:
        storageClassName: standard
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 5Gi
    config:
      cassandraYaml:
        auto_snapshot: false
        memtable_flush_writers: 1
        commitlog_segment_size_in_mb: 2
        concurrent_compactors: 1
        compaction_throughput_mb_per_sec: 0
        sstable_preemptive_open_interval_in_mb: 0
        key_cache_size_in_mb: 0
        thrift_prepared_statements_cache_size_mb: 1
        prepared_statements_cache_size_mb: 1
        slow_query_log_timeout_in_ms: 0
        counter_cache_size_in_mb: 0
        concurrent_reads: 2
        concurrent_writes: 2
        concurrent_counter_writes: 2
      jvmOptions:
        heapSize: 512Mi
    networking:
      hostNetwork: true
    resources:
      limits:
        memory: 1024Mi
    datacenters:
      - metadata:
          name: dc1
        k8sContext: kind-k8ssandra-0
        size: 1
      - metadata:
          name: dc2
        k8sContext: kind-k8ssandra-1
        size: 1
  stargate:
    size: 1
    heapSize: 512Mi
    resources:
      limits:
        memory: 1024Mi
  reaper:
    autoScheduling:
      enabled: true
```
And as expected the schema checks for Stargate fail. Here is the error from Cassandra's logs:

```
INFO  [epollEventLoopGroup-7-3] 2022-02-04 03:52:22,178 RpcExecutionException.java:39 - Failed to execute method NodeOps.createKeyspace
org.apache.cassandra.exceptions.ConfigurationException: Unrecognized strategy option {dc2} passed to NetworkTopologyStrategy for keyspace data_endpoint_auth
```